pkg/router/controller/host_admitter.go

```diff
@@ -0,0 +1,207 @@
+			// From: https://github.com/golang/go/wiki/SliceTricks
+			copy(m[idx+1:], m[idx:])
+			m[idx] = route
+			srm[key] = m
+
+			// Ensure we return from here as we change the iterator.
+			return
+		}
+	}
+
+	// Newest route or empty slice, add to the end.
+	srm[key] = append(m, route)
+}
+
+// HostAdmitter implements the router.Plugin interface to add admission
+// control checks for routes in template based, backend-agnostic routers.
+type HostAdmitter struct {
+	// plugin is the next plugin in the chain.
+	plugin router.Plugin
+
+	// admitter is a route admission function used to determine whether
+	// or not to admit routes.
+	admitter RouteAdmissionFunc
+
+	// recorder is an interface for indicating route rejections.
+	recorder RejectionRecorder
+
+	// restrictOwnership adds admission checks to restrict ownership
+	// (of subdomains) to a single owner/namespace.
+	restrictOwnership bool
+
+	// subdomainToRoute contains all routes associated with a subdomain
+	// (includes fully qualified and wildcard routes).
+	subdomainToRoute SubdomainToRouteMap
+}
+
+// NewHostAdmitter creates a plugin wrapper that checks whether or not to
+// admit routes and relay them to the next plugin in the chain.
+// Recorder is an interface for indicating why a route was rejected.
+func NewHostAdmitter(plugin router.Plugin, fn RouteAdmissionFunc, restrict bool, recorder RejectionRecorder) *HostAdmitter {
+	return &HostAdmitter{
+		plugin:   plugin,
+		admitter: fn,
+		recorder: recorder,
+
+		restrictOwnership: restrict,
+		subdomainToRoute:  make(SubdomainToRouteMap),
+	}
+}
+
+// HandleNode processes watch events on the Node resource.
+func (p *HostAdmitter) HandleNode(eventType watch.EventType, node *kapi.Node) error {
+	return p.plugin.HandleNode(eventType, node)
+}
+
+// HandleEndpoints processes watch events on the Endpoints resource.
+func (p *HostAdmitter) HandleEndpoints(eventType watch.EventType, endpoints *kapi.Endpoints) error {
+	return p.plugin.HandleEndpoints(eventType, endpoints)
+}
+
+// HandleRoute processes watch events on the Route resource.
+func (p *HostAdmitter) HandleRoute(eventType watch.EventType, route *routeapi.Route) error {
+	err, allow := p.admitter(route)
+	if !allow {
+		msg := "blocked by admitter"
+		if err != nil {
+			msg = err.Error()
+		}
+
+		glog.Errorf("Route %s not admitted: %s", routeNameKey(route), msg)
+		p.recorder.RecordRouteRejection(route, "RouteNotAdmitted", msg)
+		return err
+	}
+
+	if err != nil {
+		msg := err.Error()
+		glog.Warningf("Route %s admitted with warnings: %s", routeNameKey(route), msg)
+		p.recorder.RecordRouteRejection(route, "RouteAdmissionWarning", msg)
+	}
+
+	if p.restrictOwnership && len(route.Spec.Host) > 0 {
+		switch eventType {
+		case watch.Added, watch.Modified:
+			if err := p.addRoute(route); err != nil {
+				glog.Errorf("Route %s not admitted: %s", routeNameKey(route), err.Error())
+				p.recorder.RecordRouteRejection(route, "SubdomainAlreadyClaimed", err.Error())
```

Comment:
I'd probably switch this "HostAlreadyClaimed" or something, since it's not necessarily dealing with subdomains